runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp

```diff
@@ -199,10 +203,282 @@ void J9::ARM64::JNILinkage::restoreJNICallOutFrame(TR::Node *callNode, bool tear
+               // numIntegerArgs >= properties.getNumIntArgRegs()
+               if (childType == TR::Address || childType == TR::Int64)
+                  {
+                  op = TR::InstOpCode::strpostx;
+                  }
+               else
+                  {
+                  op = TR::InstOpCode::strpostw;
+                  }
+               mref = getOutgoingArgumentMemRef(argMemReg, argRegister, op, pushToMemory[argIndex++]);
+               argSize += 8; // always 8-byte aligned
+               }
+            numIntegerArgs++;
+            break;
+
+         case TR::Float:
+         case TR::Double:
+            if (childType == TR::Float)
+               argRegister = pushFloatArg(child);
+            else
+               argRegister = pushDoubleArg(child);
+
+            if (numFloatArgs < properties.getNumFloatArgRegs())
+               {
+               if (!cg()->canClobberNodesRegister(child, 0))
+                  {
+                  tempReg = cg()->allocateRegister(TR_FPR);
+                  op = (childType == TR::Float) ? TR::InstOpCode::fmovs : TR::InstOpCode::fmovd;
+                  generateTrg1Src1Instruction(cg(), op, callNode, tempReg, argRegister);
+                  argRegister = tempReg;
+                  }
+               if ((numFloatArgs == 0 && resType.isFloatingPoint()))
+                  {
+                  TR::Register *resultReg;
+                  if (resType.getDataType() == TR::Float)
+                     resultReg = cg()->allocateSinglePrecisionRegister();
+                  else
+                     resultReg = cg()->allocateRegister(TR_FPR);
+
+                  dependencies->addPreCondition(argRegister, TR::RealRegister::v0);
+                  dependencies->addPostCondition(resultReg, TR::RealRegister::v0);
+                  }
+               else
+                  {
+                  TR::addDependency(dependencies, argRegister, properties.getFloatArgumentRegister(numFloatArgs), TR_FPR, cg());
+                  }
+               }
+            else
+               {
+               // numFloatArgs >= properties.getNumFloatArgRegs()
+               if (childType == TR::Double)
+                  {
+                  op = TR::InstOpCode::vstrpostd;
+                  }
+               else
+                  {
+                  op = TR::InstOpCode::vstrposts;
+                  }
+               mref = getOutgoingArgumentMemRef(argMemReg, argRegister, op, pushToMemory[argIndex++]);
+               argSize += 8; // always 8-byte aligned
+               }
+            numFloatArgs++;
+            break;
+         } // end of switch
+      } // end of for
+
+   for (int32_t i = TR::RealRegister::FirstGPR; i <= TR::RealRegister::LastGPR; ++i)
+      {
+      TR::RealRegister::RegNum realReg = (TR::RealRegister::RegNum)i;
+      if (getProperties().getRegisterFlags(realReg) & ARM64_Reserved)
+         {
+         continue;
+         }
+      if (properties.getPreserved(realReg))
```

Comment:
You use both `getProperties()` (= PrivateLinkage properties) and `properties` from `_systemLinkage` here.  Is it intentional?

Reply:
Yes, it is intentional. At line 417, I use the private linkage properties because I do not want to add reserved registers such as vmthread and javastack registers to register dependencies.
The system linkage properties is used at line 421 because this if block handles registers preserved for JNI method invocation.